pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1879,33 +1924,40 @@ impl<'a> DatadirModification<'a> {
     }
 
-    // Internal helper functions to batch the modifications
-
+    // Read a page from the Timeline we are writing to.  For metadata pages, this passes through
+    // a cache in Self, which makes writes earlier in this modification visible to WAL records later
+    // in the modification.
     async fn get(&self, key: Key, ctx: &RequestContext) -> Result<Bytes, PageReconstructError> {
-        // Have we already updated the same key? Read the latest pending updated
-        // version in that case.
-        //
-        // Note: we don't check pending_deletions. It is an error to request a
-        // value that has been removed, deletion only avoids leaking storage.
-        if let Some(values) = self.pending_updates.get(&key) {
-            if let Some((_, _, value)) = values.last() {
-                return if let Value::Image(img) = value {
-                    Ok(img.clone())
-                } else {
-                    // Currently, we never need to read back a WAL record that we
-                    // inserted in the same "transaction". All the metadata updates
-                    // work directly with Images, and we never need to read actual
-                    // data pages. We could handle this if we had to, by calling
-                    // the walredo manager, but let's keep it simple for now.
-                    Err(PageReconstructError::Other(anyhow::anyhow!(
-                        "unexpected pending WAL record"
-                    )))
-                };
+        if !Self::is_data_key(&key) {
+            // Have we already updated the same key? Read the latest pending updated
+            // version in that case.
+            //
+            // Note: we don't check pending_deletions. It is an error to request a
+            // value that has been removed, deletion only avoids leaking storage.
+            if let Some(values) = self.pending_metadata_pages.get(&key.to_compact()) {
+                if let Some((_, _, value)) = values.last() {
+                    return if let Value::Image(img) = value {
+                        Ok(img.clone())
+                    } else {
+                        // Currently, we never need to read back a WAL record that we
+                        // inserted in the same "transaction". All the metadata updates
+                        // work directly with Images, and we never need to read actual
+                        // data pages. We could handle this if we had to, by calling
+                        // the walredo manager, but let's keep it simple for now.
+                        Err(PageReconstructError::Other(anyhow::anyhow!(
+                            "unexpected pending WAL record"
+                        )))
+                    };
+                }
             }
         }
+
+        // Metadata page cache miss, or we're reading a data page.
         let lsn = Lsn::max(self.tline.get_last_record_lsn(), self.lsn);
         self.tline.get(key, lsn, ctx).await
```

Comment:
Maybe I'm mis-remembering the invariants we discussed in the call about these refactorings.

But, can we assert here that `self.pending_data_pages` doesn't have an entry for `key`?

If not, why are we allowed to bypass `self.pending_data_pages`? Either way, warrants a mention in this method's doc comment. 

Also please consider `self.pending_zero_data_pages` in mentions / assertions that are added, same concern applies to them.

Reply:
> can we assert here that self.pending_data_pages doesn't have an entry for key?

We can indeed: 4e9cd3c58 -- initially I had skipped this because it's an O(N) thing, but on reflection we only hit this path during legacy db creates, so a debug-mode assertion is fine here.  Also asserting on pending_zero_data_pages at the same time.